WebUI actions

mbrysa opened the pull request:
Closes #436.

How to test:
* `# dnf install redis`
* `# systemctl start redis`
* Build and install FAF including `faf-celery-tasks` and `faf-blueprint-celery-tasks`
* `# sudo -u faf faf-migrate-db`
* `# systemctl start faf-celery-worker` (the worker)
* `# systemctl start faf-celery-beat` (the scheduler)
* Set `everyone_is_admin = true` in `/etc/faf/plugins/web2.conf`
* Restart `httpd`
* Go to `http://localhost/faf2/celery_tasks/` (or wherever FAF is running)

Alternatively use `zeromq` instead of Redis. In that case `/etc/faf/plugins/celery_tasks.conf` needs to be reconfigured.

sorki commented on config/celery-beat-env.conf:
```diff
@@ -0,0 +1,6 @@
+CELERY_APP="pyfaf.celery_tasks.celery_app"
+CELERYD_NODES="beat"
+CELERYD_OPTS="-C -S pyfaf.celery_tasks.schedulers.DBScheduler"
+CELERYD_PID_FILE="/tmp/faf-celery-beat.pid"
```
pid files should end up in /var/run

sorki replied:
https://fedoraproject.org/wiki/Packaging:Tmpfiles.d

sorki commented on config/celery-worker-env.conf:
```diff
@@ -0,0 +1,6 @@
+CELERY_APP="pyfaf.celery_tasks.celery_app"
+CELERYD_NODES="worker"
+CELERYD_OPTS="-C"
+CELERYD_PID_FILE="/tmp/faf-celery-%n.pid"
```
same here

sorki commented on faf.spec.in:
```diff
@@ -417,6 +425,13 @@ Requires: %{name}-fedmsg = %{faf_version}
+Requires: %{name} = %{faf_version}
+
+%description action-cleanup-task-results
+A plugin for %{name} implementing cleaup of old task results.
```
typo cleaup / cleanup

sorki commented on init-scripts/faf-celery-beat.service:
```diff
@@ -0,0 +1,18 @@
+Group=faf
+EnvironmentFile=/etc/faf/celery-beat-env.conf
+WorkingDirectory=/etc/faf
+ExecStart=/usr/bin/python2 -m celery beat \
```
Is there a reason to use python2 instead of just python?

sorki commented on src/pyfaf/celery_tasks/__init__.py:
```diff
@@ -0,0 +1,75 @@
+import sys
+import logging
+import datetime
+from cStringIO import StringIO
```
standard lib imports first, then 3d party libs, then pyfaf imports

sorki commented on src/pyfaf/celery_tasks/__init__.py:
```diff
@@ -0,0 +1,75 @@
+    cmdline = bunch.Bunch(params)
+
+    # Logger and fake stdout to capture the output of the action
+    my_stdout = StringIO()
```
Please use capture_output context manager:

```
from pyfaf.utils.contextmanager import captured_output
with captured_output() as (cap_stdout, cap_stderr):
    <code>
```

cap_std* are StringIO objects, sys.stdout/err is saved/restored automatically.

sorki commented:
- Needs documentation 
- Needs ansible support
- automatically generated field names and help texts should have first letter upcased
- list of available actions is too cluttered

Otherwise very nice!

sorki commented:
Using */5 as minutes doesn't work. Not a big deal but it would be nice to have this shorthand.


> In vixie-cron, slashes can be combined with ranges to specify step values. For example, */7 in the 
> minutes field indicates every 7 minutes (see note). It is shorthand for the more verbose POSIX form 
> 5,10,15,20,25,30,35,40,45,50,55,00. POSIX does not define a use for slashes; its rationale 
> (commenting on a BSD extension) notes that the definition is based on System V format but does not exclude the possibility of extensions.

mbrysa commented:
Rebased:
* pid files in `/run/faf-celery/` using tmpfiles.d
* typo fixed
* python2 -> python
* imports reordered
* added and using `capture_output_combined` context manager
* actions list split into three columns

Using */5 as minutes works for me. Could you please retest?

sorki commented on init-scripts/faf-celery-beat.service:
```diff
@@ -0,0 +1,18 @@
+ExecStart=/usr/bin/python -m celery beat \
+    -A $CELERY_APP --pidfile=${CELERYD_PID_FILE} \
+    --logfile=${CELERYD_LOG_FILE} --loglevel=${CELERYD_LOG_LEVEL} \
+    --detach $CELERYD_OPTS
```
Only works for me without --detach and without Type=forking (both are redundant)

sorki commented on faf.spec.in:
```diff
@@ -1133,6 +1208,22 @@ fi
+%{_unitdir}/faf-celery-worker.service
+%config(noreplace) %{_sysconfdir}/faf/celery-beat-env.conf
+%config(noreplace) %{_sysconfdir}/faf/celery-worker-env.conf
+%{_tmpfilesdir}/faf-celery-tmpfiles.conf
```
Also

`%dir /run/faf-celery/`

mbrysa commented:
Forking removed, `/run/faf-celery` dir added.
